### PR TITLE
9.0 alter multiple subscription

### PR DESCRIPTION
**Pull request by robinkeunen**
[ref: Plusieurs abonnements de plusieurs magazines](https://gestion.coopiteasy.be/web#id=2190&view_type=form&model=project.task&action=479)

- [x] 1. add subscription template on subscription object
- [x] 2. each request generates a new subscription (through invoices)
- [x] 3. check product release flow
- [x] 4. refactor simplified processes due to 2

**Only review last 2 commits, the rest is cosmetics**
- [ADD] ps: subscription template on subscription object 213bc97
- [REF] ps: each request generates a new subscription (through invoices) 70b17d0

### CODE REVIEW

File: product_subscription/__openerp__.py [hunk 1]
```diff
@@ -1,41 +1,28 @@
...
-#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-#    GNU Affero General Public License for more details.
-#
-#    You should have received a copy of the GNU Affero General Public License
-#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
-#
-##############################################################################
+# Copyright 2019 Coop IT Easy SCRL fs
+#   Houssine Bakkali <houssine@coopiteasy.be>
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
```
Comment (houssine78) [inline product_subscription/__openerp__.py:5]: @robinkeunen you may add your name here now

File: product_subscription/__openerp__.py [hunk 2]
```diff
@@ -1,41 +1,28 @@
...
+        'account_analytic_distribution',
+        'l10n_be_invoice_bba',  # todo delocalize
+        'email_template_config',
+        'mail',
     ],
-    "author": "Houssine BAKKALI <houssine@coopiteasy.be>",
-    "category": "Sales",
-    "website": "www.coopiteasy.be",
-    "description": """
+    'author': 'Houssine BAKKALI <houssine@coopiteasy.be>',
```
Comment (houssine78) [inline product_subscription/__openerp__.py]: we didn't decided to put coop it easy as author ?

File: product_subscription/models/invoice.py [hunk 3]
```diff
@@ -1,50 +1,55 @@
...
-                    'payment_date': effective_date}
+        self.ensure_one()
 
-        sub_req = self.product_subscription_request
-        sub_template = sub_req.subscription_template
-        # check if there is already an ongoing or an old subscription
-        # tied to the subscriber
+        subscription_template = (
+            self.product_subscription_request.subscription_template
+        )
```
Comment (houssine78) [inline product_subscription/models/invoice.py:27]: that pep8 stuff go to far :)

File: product_subscription/models/invoice.py [hunk 4]
```diff
@@ -1,50 +1,55 @@
 # -*- coding: utf-8 -*-
+# Copyright 2019 Coop IT Easy SCRL fs
+#   Houssine Bakkali <houssine@coopiteasy.be>
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
```
Comment (houssine78) [inline product_subscription/models/invoice.py:5]: add your name here

File: product_subscription/models/product_release.py [hunk 5]
```diff
@@ -1,49 +1,95 @@
 # -*- coding: utf-8 -*-
-from openerp import models, fields, api
+# Copyright 2019 Coop IT Easy SCRL fs
+#   Houssine Bakkali <houssine@coopiteasy.be>
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
```
Comment (houssine78) [inline product_subscription/models/product_release.py:5]: add your name here

**Review by houssine78 — APPROVED** [review]

File: product_subscription/__openerp__.py (hunk 1 above)
Comment (houssine78) [reply product_subscription/__openerp__.py:5]: ahah

File: product_subscription/__openerp__.py (hunk 2 above)
Comment (robinkeunen) [reply product_subscription/__openerp__.py]: :+1:

File: product_subscription/models/invoice.py (hunk 3 above)
Comment (robinkeunen) [reply product_subscription/models/invoice.py:27]: Jme suis détendu par rapport à pep8 mais je préfère comme ça :-)

File: product_subscription/models/product_release.py [hunk 6]
```diff
@@ -91,6 +91,7 @@ def action_view_delivery(self):
         of given sales order ids. It can either be a in a list or in a form
         view, if there is only one delivery order to show.
         """
+        # todo move to ps_delivery?
```
Comment (houssine78) [inline product_subscription/models/product_release.py:94]: what is ps_delivery ?

File: product_subscription/models/product_release.py [hunk 7]
```diff
@@ -115,89 +116,86 @@ def action_view_delivery(self):
...
+        subscriptions = self.env['product.subscription.object'].search([
+            ('counter', '>', 0),
+            ('subscription_template.released_products', 'in', self.product_id.id),
+        ])
 
         for subscription in subscriptions:
-            line_vals['subscriber'] = subscription.subscriber.id
-            line_vals['product_subscription'] = subscription.id
-            release_line_obj.create(line_vals)
-        self.write(vals)
```
Comment (houssine78) [inline product_subscription/models/product_release.py]: where this has been moved ?

File: product_subscription/models/product_release.py [hunk 8]
```diff
@@ -115,89 +116,86 @@ def action_view_delivery(self):
...
 
-        vals = {'state': 'validated'}
         if self.name == '' or not self.name:
             prod_rel_seq = self.env.ref('product_subscription.sequence_product_release', False)  # noqa
-            vals['name'] = prod_rel_seq.next_by_id()
-
-        subscriptions = prod_sub_obj.search([('counter', '>', 0)])
+            name = prod_rel_seq.next_by_id()
+        else:
+            name = self.name
```
Comment (houssine78) [inline product_subscription/models/product_release.py:141]: why setting name to its name ? I prefer doing it with vals

**Review by houssine78 — APPROVED** [review]

File: product_subscription/models/product_release.py (hunk 6 above)
Comment (robinkeunen) [reply product_subscription/models/product_release.py:94]: `product_subscription_delivery`

File: product_subscription/models/product_release.py (hunk 8 above)
Comment (robinkeunen) [reply product_subscription/models/product_release.py:141]: That's what I do at line 156. I need to set the `name` variable in the else to make sure it's in the scope.

Comment (houssine78) [reply product_subscription/models/product_release.py:141]: This is why vals was for at the first place. you only set name in vals dict if needed and you avoid an useless else statement. As state will always be 'validated' setting name is just for the case where it has no name already.

**Review by houssine78 — CHANGES_REQUESTED** [review]
make some tests taking into account the performances. trigger a change on the field for each counter field update could lead to some perf issue.

**Pull request closed without merging**